File: Thisaravi-Backend/parsers.py
import re
import json


def extract_json(text: str) -> dict:
    """Extracts JSON from markdown code blocks or raw strings"""
    try:
        # Try finding markdown JSON block
        match = re.search(r'```json\s*(.*?)\s*```', text, re.DOTALL)
        if match: return json.loads(match.group(1))
        
        # Try finding raw JSON object
        match = re.search(r'(\{.*\})', text, re.DOTALL)
        if match: return json.loads(match.group(1))
        
        # Try regular load
        return json.loads(text)
    except Exception:
        return {}


def parse_structured_text(text: str) -> dict:
    """
    Parses the raw text output from the fine-tuned Gemma model into structured JSON.
    
    Expected Text Format:
    Gap Analysis: <text> 
    Match Score: <number>% 
    Project Recommendation: <title> 
    Tech Stack: <list> 
    Steps: <list>
    """
    try:
        # Regex Extraction - Fuzzy / Robust
        # 1. Gap Analysis: Look for "Gap Analysis" or "Analysis" header
        # Followed by colon/space OR newline
        gap_match = re.search(r'(?:Gap Analysis|Analysis)(?:[:\s-]*|\s*\n)(.*?)(?=(?:\*\*|#|Match Score|Project)|$)', text, re.DOTALL | re.IGNORECASE)
        
        # 2. Match Score: Look for "Match Score" or just a percentage near the top if specific header missing
        score_match = re.search(r'Match Score[^\d]*(\d+)', text, re.IGNORECASE)
        
        # 3. Project Rec: Look for "Project Recommendation" etc.
        proj_match = re.search(r'(?:Project Recommendation|Recommended Project|Project)(?:[:\s-]*|\s*\n)(.*?)(?=(?:\*\*|#|Tech Stack)|$)', text, re.DOTALL | re.IGNORECASE)
        
        # 4. Tech Stack: "Tech Stack"
        stack_match = re.search(r'(?:Tech Stack|Technologies|Stack)(?:[:\s-]*|\s*\n)(.*?)(?=(?:\*\*|#|Steps|Implementation)|$)', text, re.DOTALL | re.IGNORECASE)
        
        # 5. Steps: "Steps"
        steps_match = re.search(r'(?:Steps|Implementation|Plan)(?:[:\s-]*|\s*\n)(.*)', text, re.DOTALL | re.IGNORECASE)

        # Fallback for Gap Analysis if regex failed but we have text
        gap_text = gap_match.group(1).strip() if gap_match else (text[:500] + "..." if len(text) > 500 else text)

        # Parsing Lists (Tech Stack & Steps)
        raw_stack = stack_match.group(1).strip() if stack_match else ""
        tech_stack = [s.strip() for s in re.split(r',|and|\n|-|\*|;', raw_stack) if s.strip()]
        
        raw_steps = steps_match.group(1).strip() if steps_match else ""
        steps = [s.strip() for s in re.split(r'\d+\.|-|\*', raw_steps) if s.strip()]

        # Construct JSON even if some matches failed (Best Effort)
        data = {
            "gap_analysis": {
                "analysis_summary": gap_text,
                "match_percentage": int(score_match.group(1)) if score_match else 0,
                "missing_skills": [] 
            },
            "project_recommendation": {
                "project_title": proj_match.group(1).strip() if proj_match else "Recommended Capstone",
                "objective": "Capstone Project",
                "tech_stack": tech_stack if tech_stack else ["Python", "General Dev"],
                "implementation_steps": steps if steps else ["Review full analysis for details."]
            }
        }
        return data
            
    except Exception as e:
        print(f"Parser Exception: {e}")
        return {"error": f"Failed to parse text: {str(e)}", "raw_text": text}


def parse_response(text: str) -> dict:
    """
    Master parsing function that attempts to handle both JSON and Structured Text formats.
    """
    # 1. Attempt JSON Parsing first (Generic Models)
    json_data = extract_json(text)
    if json_data and "gap_analysis" in json_data:
        return json_data
        
    # 2. Attempt Structured Text Parsing (Fine-tuned Model)
    text_data = parse_structured_text(text)
    if text_data and "gap_analysis" in text_data:
        return text_data
        
    # 3. Fallback / Error
    return {
        "error": "Could not parse output",
        "raw_text": text
    }

File: Thisaravi-frontend/src/lib/parsers.ts
import type { AnalysisResult } from '@/types/api';

export function extractJson(text: string): Record<string, unknown> {
  try {
    // Try markdown JSON block
    const mdMatch = text.match(/```json\s*([\s\S]*?)\s*```/);
    if (mdMatch) return JSON.parse(mdMatch[1]);

    // Try raw JSON object
    const rawMatch = text.match(/(\{[\s\S]*\})/);
    if (rawMatch) return JSON.parse(rawMatch[1]);

    // Try direct parse
    return JSON.parse(text);
  } catch {
    return {};
  }
}

export function parseStructuredText(text: string): AnalysisResult {
  try {
    // 1. Gap Analysis
    const gapMatch = text.match(
      /(?:Gap Analysis|Analysis)(?:[:\s-]*|\s*\n)([\s\S]*?)(?=(?:\*\*|#|Match Score|Project)|$)/i
    );
    // 2. Match Score
    const scoreMatch = text.match(/Match Score[^\d]*(\d+)/i);
    // 3. Project Recommendation
    const projMatch = text.match(
      /(?:Project Recommendation|Recommended Project|Project)(?:[:\s-]*|\s*\n)([\s\S]*?)(?=(?:\*\*|#|Tech Stack)|$)/i
    );
    // 4. Tech Stack
    const stackMatch = text.match(
      /(?:Tech Stack|Technologies|Stack)(?:[:\s-]*|\s*\n)([\s\S]*?)(?=(?:\*\*|#|Steps|Implementation)|$)/i
    );
    // 5. Steps
    const stepsMatch = text.match(
      /(?:Steps|Implementation|Plan)(?:[:\s-]*|\s*\n)([\s\S]*)/i
    );

    const gapText = gapMatch
      ? gapMatch[1].trim()
      : text.length > 500
        ? text.slice(0, 500) + '...'
        : text;

    const rawStack = stackMatch ? stackMatch[1].trim() : '';
    const techStack = rawStack
      // Split only on commas, "and", semicolons, or a newline followed by a bullet/dash
      // Do NOT split on bare "-" — that breaks hyphenated tech names like "CI/CD"
      .split(/,|;|\band\b|\n\s*[-*]\s*/)
      .map((s) => s.replace(/^[-*\s]+/, '').trim())
      .filter(Boolean);

    const rawSteps = stepsMatch ? stepsMatch[1].trim() : '';
    // Only split on actual list markers at the start of a line:
    //   - "1." / "1)" / "1:" followed by whitespace
    //   - "Step N" variants
    //   - "- " or "* " (dash/bullet followed by a space — NOT bare "-" inside words)
    // This prevents fragmentation of hyphenated words like "domain-specific".
    const stepsNormalized = rawSteps.replace(/\r\n/g, '\n');
    const steps = stepsNormalized
      .split(/\n\s*(?:\d+[.):\s]\s+|Step\s+\d+[.:)?\s]\s*|[-*]\s+)/)
      .map((s) => s.replace(/^(?:Step\s*)?\d+[.):\s]\s*/i, '').trim())
      .filter((s) => s.length > 5);

    return {
      gap_analysis: {
        analysis_summary: gapText,
        match_percentage: scoreMatch ? parseInt(scoreMatch[1], 10) : 0,
        missing_skills: [],
      },
      project_recommendation: {
        project_title: projMatch ? projMatch[1].trim() : 'Recommended Capstone',
        objective: 'Capstone Project',
        tech_stack: techStack.length ? techStack : ['Python', 'General Dev'],
        implementation_steps: steps.length ? steps : ['Review full analysis for details.'],
      },
    };
  } catch (e) {
    return {
      gap_analysis: { analysis_summary: '', match_percentage: 0, missing_skills: [] },
      project_recommendation: { project_title: '', objective: '', tech_stack: [], implementation_steps: [] },
      error: `Failed to parse text: ${e}`,
      raw_text: text,
    };
  }
}

export function parseResponse(text: string): AnalysisResult {
  // 1. Try JSON first (generic models)
  const jsonData = extractJson(text);
  if (jsonData && 'gap_analysis' in jsonData) {
    const result = jsonData as unknown as AnalysisResult;
    // Coerce match_percentage to a number — some models return "85%" or "85"
    const raw = (result.gap_analysis as Record<string, unknown>).match_percentage;
    if (typeof raw !== 'number') {
      (result.gap_analysis as Record<string, unknown>).match_percentage =
        parseInt(String(raw).replace(/[^\d]/g, ''), 10) || 0;
    }
    return result;
  }

  // 2. Try structured text parsing (fine-tuned model)
  const textData = parseStructuredText(text);
  if (textData && 'gap_analysis' in textData && !textData.error) {
    return textData;
  }

  // 3. Fallback
  return {
    gap_analysis: { analysis_summary: '', match_percentage: 0, missing_skills: [] },
    project_recommendation: { project_title: '', objective: '', tech_stack: [], implementation_steps: [] },
    error: 'Could not parse output',
    raw_text: text,
  };
}
